### PR TITLE
Andrew/control

### PR DESCRIPTION
**Pull request by exspeed**

### CODE REVIEW

File: src/Camera.cpp [hunk 1]
```diff
@@ -1,29 +1,58 @@
...
 
 Camera::Camera()
 	: ProjectionMatrix(glm::mat4(1.0))
 	, ViewMatrix(glm::mat4(1.0))
 {
 	// Initial position
-	glm::vec3 position = glm::vec3(4,3,3);
+	pos = glm::vec3(4,3,3);
+	forward = glm::vec3(0,0,0);
+	up = glm::vec3(0,1,0);
```
Comment (ammunoz) [inline src/Camera.cpp]: These are better off initialized in the initializer list.

File: src/Camera.cpp [hunk 2]
```diff
@@ -1,29 +1,58 @@
...
+
 	// Initial Field of View
 	float initialFoV = 45.0f;
 
 	ViewMatrix = glm::lookAt(
```
Comment (ammunoz) [inline src/Camera.cpp]: Left bracket should move to next line to keep consistent with rest of the code.

File: src/Camera.cpp [hunk 3]
```diff
@@ -1,29 +1,58 @@
...
+		up  // Head is up (set to 0,-1,0 to look upside-down)
 	);
 
 	//perspective (T const &fovy, T const &aspect, T const &near, T const &far)
 	ProjectionMatrix = glm::perspective(glm::radians(initialFoV), 4.0f / 3.0f, 0.1f, 100.0f);
 }
 
+void Camera::UpdateViewMatrix()
+{
+	ViewMatrix = glm::lookAt(
```
Comment (ammunoz) [inline src/Camera.cpp]: Left bracket should move to next line to keep consistent with rest of the code.

File: src/Camera.cpp [hunk 4]
```diff
@@ -1,29 +1,58 @@
...
+
+void Camera::RotateAround(float deltaX, float deltaY)
+{
+	// look left and right
+	glm::mat4 yaw = glm::rotate(deltaX *0.5f, up);
+	forward = yaw*(glm::vec4(forward-pos, 0));
+
+	// look up and down
+	glm::vec3 bi = normalize(cross(up, (forward-pos)));
+	glm::mat4 roll = glm::rotate(-deltaY *0.5f, bi);
```
Comment (ammunoz) [inline src/Camera.cpp]: Why 0.5f? Please name it a constant so we know where it came from and can change it easily.

File: src/GLHelpers.cpp [hunk 5]
```diff
@@ -87,9 +84,22 @@ void ErrorCallback(int error, const char* description)
...
-// Handles keyboard input events
-void KeyCallback(GLFWwindow* window, int key, int scancode, int action, int mods)
+
+
+// GLM DEBUG
+void PrintMat4(glm::mat4 const &m)
+{
+	for(int i = 0; i < 4; i++)
+	{
+		for(int j = 0; j< 4; j++)
```
Comment (ammunoz) [inline src/GLHelpers.cpp]: [Not important] minor inconsistent spacing between j and <

File: src/InputManager.cpp [hunk 6]
```diff
@@ -0,0 +1,34 @@
+#include "InputManager.h"
+
+InputManager::InputManager(GLFWwindow *w, Camera* cam)
+:window(w)
+,camera(cam)
+{
+}
+
+void InputManager::CheckInput(){
```
Comment (ammunoz) [inline src/InputManager.cpp]: Left bracket should move to next line to keep consistent with rest of the code.

File: src/InputManager.cpp [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+	KeyInput();
+	MouseInput();
+}
+
+void InputManager::MouseInput() {
+	double xpos, ypos;
+	if(glfwGetMouseButton(window, GLFW_MOUSE_BUTTON_LEFT) == GLFW_PRESS){
```
Comment (ammunoz) [inline src/InputManager.cpp]: Left bracket should move to next line to keep consistent with rest of the code.

File: src/InputManager.cpp [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+		float deltaY = (ypos - cursorY) * 0.01;
+		camera->RotateAround(deltaX, deltaY);
+	}
+	
+	glfwGetCursorPos(window, &xpos, &ypos);
+	cursorX = xpos;
+	cursorY = ypos;
+	
+}
+void InputManager::KeyInput() {
```
Comment (ammunoz) [inline src/InputManager.cpp]: Left bracket should move to next line to keep consistent with rest of the code.

File: src/InputManager.cpp [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+	MouseInput();
+}
+
+void InputManager::MouseInput() {
+	double xpos, ypos;
+	if(glfwGetMouseButton(window, GLFW_MOUSE_BUTTON_LEFT) == GLFW_PRESS){
+		// Get mouse position
+		glfwGetCursorPos(window, &xpos, &ypos);
+		float deltaX = (xpos - cursorX) * 0.01;
+		float deltaY = (ypos - cursorY) * 0.01;
```
Comment (ammunoz) [inline src/InputManager.cpp]: Why 0.01? Please name it a constant so we know where it came from and can change it easily.

Also, in c++, 0.01 is a double which then has to unnecessarily cast to float before multiplication with another float.

File: src/InputManager.cpp [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+	}
+	
+	glfwGetCursorPos(window, &xpos, &ypos);
+	cursorX = xpos;
+	cursorY = ypos;
+	
+}
+void InputManager::KeyInput() {
+	
+	if(glfwGetKey(window, GLFW_KEY_ESCAPE) == GLFW_PRESS){
```
Comment (ammunoz) [inline src/InputManager.cpp]: Left bracket should move to next line to keep consistent with rest of the code.

File: src/main.cpp [hunk 11]
```diff
@@ -49,6 +54,7 @@ int main(int argc, char *argv[])
 
 	// Clean up allocated resources before exit
 	scene.ClearGeometries();
+	delete(camera);
```
Comment (ammunoz) [inline src/main.cpp:58]: Nice

File: src/Camera.cpp [hunk 12]
```diff
@@ -1,29 +1,58 @@
...
-		position, // Camera position in World Space
-		glm::vec3(0,0,0), // and looks at the origin
-		glm::vec3(0,1,0)  // Head is up (set to 0,-1,0 to look upside-down)
+		pos, // Camera position in World Space
+		forward, // and looks at the origin
+		up  // Head is up (set to 0,-1,0 to look upside-down)
 	);
 
 	//perspective (T const &fovy, T const &aspect, T const &near, T const &far)
 	ProjectionMatrix = glm::perspective(glm::radians(initialFoV), 4.0f / 3.0f, 0.1f, 100.0f);
```
Comment (ammunoz) [inline src/Camera.cpp]: Please name magic numbers as consts so we know where they came from and can change them easily.

File: include/InputManager.h [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+	
+	InputManager(GLFWwindow *window, Camera* cam);
+	void CheckInput();
+	
+private:
+	float cursorX = 0;
+	float cursorY = 0;
+	
+	void KeyInput();
+	void MouseInput();
```
Comment (ammunoz) [inline include/InputManager.h]: I think CheckInput(), KeyInput(), and MouseInput() can all be const functions.

**Review by ammunoz — CHANGES_REQUESTED** [review]

File: src/Camera.cpp (hunk 2 above)
Comment (exspeed) [reply src/Camera.cpp]: d

File: src/Camera.cpp (hunk 3 above)
Comment (exspeed) [reply src/Camera.cpp]: d

File: src/GLHelpers.cpp (hunk 5 above)
Comment (exspeed) [reply src/GLHelpers.cpp]: d

File: src/InputManager.cpp (hunk 10 above)
Comment (exspeed) [reply src/InputManager.cpp]: d

File: src/Camera.cpp (hunk 12 above)
Comment (exspeed) [reply src/Camera.cpp]: d

File: include/InputManager.h (hunk 13 above)
Comment (exspeed) [reply include/InputManager.h]: d

File: src/InputManager.cpp (hunk 9 above)
Comment (exspeed) [reply src/InputManager.cpp]: d

File: src/Camera.cpp (hunk 4 above)
Comment (exspeed) [reply src/Camera.cpp]: d

File: include/InputManager.h [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+	
+	InputManager(GLFWwindow *window, Camera* cam);
+	const void CheckInput();
+	
+private:
+	float cursorX = 0;
+	float cursorY = 0;
+	
+	const void KeyInput();
+	const void MouseInput();
```
Comment (ammunoz) [inline include/InputManager.h]: You probably want:
`void MouseInput() const;`
as opposed to
`const void MouseInput();`

The first makes MouseInput() a const function.
The second means that MouseInput() returns a const void.

File: include/Scene.h [hunk 15]
```diff
@@ -11,14 +11,14 @@
 class Scene
 {
 public:
-	Scene(std::string vertex_shader, std::string fragment_shader);
+	Scene(std::string vertex_shader, std::string fragment_shader,Camera* cam);
```
Comment (ammunoz) [inline include/Scene.h]: [Not important] Minor spacing issue

File: src/Camera.cpp [hunk 16]
```diff
@@ -1,29 +1,57 @@
...
+		forward, // and looks at the origin
+		up  // Head is up (set to 0,-1,0 to look upside-down)
 	);
 
 	//perspective (T const &fovy, T const &aspect, T const &near, T const &far)
-	ProjectionMatrix = glm::perspective(glm::radians(initialFoV), 4.0f / 3.0f, 0.1f, 100.0f);
+	float initialFoV = 45.0f;
+	float near = 0.1f;
+	float far = 100.f;
+	float aspect = 1.0;
```
Comment (ammunoz) [inline src/Camera.cpp:28]: [Minor issue] Can make these const variables as well. Prob better to keep these as Camera members in header.

File: src/GLHelpers.cpp [hunk 17]
```diff
@@ -15,15 +15,12 @@ void Initialize(GLFWwindow*& window)
 	glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);
 	glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
 	int width = 512, height = 512;
-	window = glfwCreateWindow(width, height, "CPSC 453 OpenGL Boilerplate", 0, 0);
+	window = glfwCreateWindow(width, height, "CPSC 589 Modelling Project", 0, 0);
```
Comment (ammunoz) [inline src/GLHelpers.cpp:18]: good catch haha

File: src/InputManager.cpp [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+const void InputManager::MouseInput()
+{
+	double xpos, ypos;
+	if(glfwGetMouseButton(window, GLFW_MOUSE_BUTTON_LEFT) == GLFW_PRESS)
+	{
+		// Get mouse position
+		glfwGetCursorPos(window, &xpos, &ypos);
+		float sensitivity = 0.005f;
```
Comment (ammunoz) [inline src/InputManager.cpp]: Prob better to keep this as member variable of InputManager as mouseSensitivity or something.

Can make it const as well.

File: src/Camera.cpp [hunk 19]
```diff
@@ -1,29 +1,57 @@
...
 
 
-
-Camera::Camera()
-	: ProjectionMatrix(glm::mat4(1.0))
-	, ViewMatrix(glm::mat4(1.0))
+Camera::Camera(glm::vec3 p, glm::vec3 f, glm::vec3 u):
+	pos(p) 
+, forward(f)
+, up(u)
```
Comment (ammunoz) [inline src/Camera.cpp:14]: Weird spacing

File: src/GLHelpers.cpp [hunk 20]
```diff
@@ -87,9 +84,22 @@ void ErrorCallback(int error, const char* description)
...
+		}
+		 std::cout << std::endl;
+	}
+}
+
+void PrintVec3(glm::vec3 const &v)
 {
-	if (key == GLFW_KEY_ESCAPE && action == GLFW_PRESS)
-		glfwSetWindowShouldClose(window, GL_TRUE);
+			std::cout << v[0] << " " << v[1] << " " << v[2] << std::endl;
```
Comment (ammunoz) [inline src/GLHelpers.cpp]: [Minor issue] spacing

File: src/InputManager.cpp [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+	
+	glfwGetCursorPos(window, &xpos, &ypos);
+	cursorX = xpos;
+	cursorY = ypos;
+	
+}
+const void InputManager::KeyInput()
+{
+	
+	if(glfwGetKey(window, GLFW_KEY_ESCAPE) == GLFW_PRESS){
```
Comment (ammunoz) [inline src/InputManager.cpp]: Left bracket should go to next line to keep consistent with the rest of the code.

**Review by ammunoz — CHANGES_REQUESTED** [review]

File: include/InputManager.h (hunk 14 above)
Comment (exspeed) [reply include/InputManager.h]: I'm setting some values. guess it can't be const
cursorX = xpos;
cursorY = ypos;

File: include/Scene.h (hunk 15 above)
Comment (exspeed) [reply include/Scene.h]: d

File: src/InputManager.cpp (hunk 18 above)
Comment (exspeed) [reply src/InputManager.cpp]: d

File: src/GLHelpers.cpp (hunk 20 above)
Comment (exspeed) [reply src/GLHelpers.cpp]: d

File: src/Camera.cpp (hunk 19 above)
Comment (exspeed) [reply src/Camera.cpp:14]: looks fine to me

**Review by ammunoz — APPROVED** [review]

**Pull request merged**